### ISSUE TITLE
Tar crashes on reading the specific file

### ISSUE THREAD

**Issue by iSazonov** [dotnet/runtime#73440]
### Description

I am experimenting with new Tar API to develop PowerShell cmdlet and catch the issue.

I get `System.IO.EndOfStreamException` if reading the specific file. Other small tar and tar.gz files I created for testing were read well.

### Reproduction Steps

1. Download https://github.com/PowerShell/PowerShell/releases/download/v7.3.0-preview.6/powershell-7.3.0-preview.6-linux-x64.tar.gz
2. Try read with simple code:
```c#
            using (var sourceStream = File.Open(LiteralPath, FileMode.Open, FileAccess.Read, FileShare.Read))
            {
                using var decompressorStream = new GZipStream(sourceStream, CompressionMode.Decompress);
                using var reader = new TarReader(decompressorStream);
                TarEntry? entry = null;
                try
                {
                    entry = reader.GetNextEntry();
                }
                catch
                {
                    // It is not tar archive. Go to next format detection.
                }

                if (entry is not null)
                {
                    do
                    {
                        // WriteObject(entry); --- PowerShell output
                    }
                    while ((entry = reader.GetNextEntry()) is not null);

                    return;
                }
            }
```

### Expected behavior

No exception.

### Actual behavior

```
Exception             :
    Type       : System.IO.EndOfStreamException
    TargetSite :
        Name          : AdvanceStream
        DeclaringType : System.Formats.Tar.TarHelpers, System.Formats.Tar, Version=7.0.0.0, Culture=neutral, PublicKeyT
oken=cc7b13ffcd2ddd51
        MemberType    : Method
        Module        : System.Formats.Tar.dll
    Message    : Attempted to read past the end of the stream.
    Source     : System.Formats.Tar
    HResult    : -2147024858
    StackTrace :
   at System.Formats.Tar.TarHelpers.AdvanceStream(Stream archiveStream, Int64 bytesToDiscard)
   at System.Formats.Tar.TarReader.AdvanceDataStreamIfNeeded()
   at System.Formats.Tar.TarReader.GetNextEntry(Boolean copyData)
   at Microsoft.PowerShell.Archive.GetArchiveEntriesCommand.EndProcessing() in C:\Users\1\Documents\GitHub\iSazonov\Mic
rosoft.PowerShell.Archive\src\GetArchiveEntriesCommand.cs:line 103
   at System.Management.Automation.Cmdlet.DoEndProcessing() in C:\Users\1\Documents\GitHub\iSazonov\PowerShell\src\Syst
em.Management.Automation\engine\cmdlet.cs:line 187
   at System.Management.Automation.CommandProcessorBase.Complete() in C:\Users\1\Documents\GitHub\iSazonov\PowerShell\s
rc\System.Management.Automation\engine\CommandProcessorBase.cs:line 593
CategoryInfo          : NotSpecified: (:) [Get-ArchiveEntries], EndOfStreamException
FullyQualifiedErrorId : System.IO.EndOfStreamException,Microsoft.PowerShell.Archive.GetArchiveEntriesCommand
InvocationInfo        :
    MyCommand        : Get-ArchiveEntries
    ScriptLineNumber : 1
    OffsetInLine     : 4
    HistoryId        : 18
    Line             : $a=Get-ArchiveEntries -LiteralPath C:\Users\1\Downloads\powershell-7.2.0-preview.3-linux-x64.tar
.gz -DestinationPath q
    PositionMessage  : At line:1 char:4
                       + $a=Get-ArchiveEntries -LiteralPath C:\Users\1\Downloads\powershell-7. …
                       +    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    InvocationName   : Get-ArchiveEntries
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Configuration

7.0.100-preview.6.22352.1

**Comment by iSazonov** [comment]
Such issue was mentioned in #68230

**Comment by carlossanlop** [comment]
@iSazonov I downloaded the file you shared: https://github.com/PowerShell/PowerShell/releases/download/v7.3.0-preview.6/powershell-7.3.0-preview.6-linux-x64.tar.gz

... but it is malformed.

I tried opening it with 7-zip and I got this error:

<img src="https://user-images.githubusercontent.com/1175054/183513045-0b0700f2-e4af-4e2f-b974-88acea38c8cb.png" width="500" />

I also tried opening it with your code snippet, but I removed the try catch, and the first call to `GetNextEntry` is throwing an exception because it is not a valid tar file, which is expected.

Can you still repro the issue with other well-formed tar.gz files?

**Comment by carlossanlop** [comment]
I downloaded it again. I don't know why, but I can open it now with 7-zip.

I tried your code snippet once again. I tested it using the latest bits in `main` in dotnet/runtime. The test passes, no exceptions are thrown, and I verified that all the entries in the tar.gz are visited in the while loop.

There were several bug fixes introduced between preview5 and preview6. Right now, main is on preview7. I suspect that is the reason why you're hitting such bug, and I am not.
